#4 react-v16.13.0.md add japanese transration (#329)
reactjs/ja.reactjs.org: content/blog/2020-02-26-react-v16.13.0.md

  - "blog/2020/03/02/react-v16.13.0.html"
---

本日、React 16.13.0 をリリースします。このリリースには、今後のメジャーリリースに備えて、バグ修正と新しい非推奨の警告が含まれています。

> ## New Warnings {#new-warnings}

Author: 了解しました。修正します。

> ## 新規注意事項 {#new-warnings}

Reviewer: ```suggestion
## 新たな警告 {#new-warnings}
```

この warnings は本当に console に出てくる warning のことです

> ###  レンダリング中のいくつかの更新に関する注意点{#warnings-for-some-updates-during-render}

Reviewer: ```suggestion
###  レンダリング中のいくつかの更新に関する警告 {#warnings-for-some-updates-during-render}
```

React コンポーネントは、レンダー中に他のコンポーネントに副作用を起こしてはいけません。

レンダー中に `setState` を呼び出すことはサポートされていますが [*同じ*コンポーネントに対してのみ](/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops)可能です。別のコンポーネントのレンダー中に `setState` を呼び出すと、警告が表示されるようになりました。

```
Warning: Cannot update a component from inside the function body of a different component.
```

**この警告は、意図しない状態変更によって引き起こされるアプリケーションのバグを見つけるのに役立ちます。**レンダーの結果として他のコンポーネントの状態を意図的に変更したいという稀なケースでは、`setState` 呼び出しを `useEffect` にラップすることができます。

> ### スタイルルールが矛盾している場合の注意点{#warnings-for-conflicting-style-rules}

Reviewer: ```suggestion
### スタイルルールが矛盾している場合の警告 {#warnings-for-conflicting-style-rules}
```

</div> 
```

この `<div>` は `toggle` の値に関係なく常に背景色が赤になると思われるかもしれません。しかし、[このデモを見てわかるように](https://codesandbox.io/s/suspicious-sunset-g3jub)、`toggle` の値を `true` と `false` の間で切り替えると、背景色は `red` から始まり、その後 `transparent` と `blue` の間で交互に切り替わります。
  

> **Reactは、スタイルルールの競合を検出し、警告をログに記録するようになりました。**この問題を修正するには、同一の CSS プロパティの短い記法と長い記法のバージョンを `style` プロパティ内で混在させないようにしてください。

Reviewer: ```suggestion
**React は、スタイルルールの競合を検出し、警告をログに記録するようになりました。**この問題を修正するには、同一の CSS プロパティの短い記法と長い記法のバージョンを `style` プロパティ内で混在させないようにしてください。
```

### 一部の非推奨 string ref に関する注意点 {#warnings-for-some-deprecated-string-refs}

> [String ref（文字列形式の ref）は古いレガシー API](/docs/refs-and-the-dom.html#legacy-api-string-refs)であり、既に勧められておらず、将来的に正式に非推奨となる予定です。

Reviewer: ```suggestion
[String ref（文字列形式の ref）は古いレガシー API](/docs/refs-and-the-dom.html#legacy-api-string-refs) であり、既に勧められておらず、将来的に正式に非推奨となる予定です。
```

```js
<Button ref="myRef" />
```

> (String ref と異なり、一般的なref自体は**完全にサポートされ続けます**ので混同しないようにしてください。）

Reviewer: ```suggestion
（String ref と異なり、一般的なref自体は**完全にサポートされ続けます**ので混同しないようにしてください）
```

> （String ref と異なり、一般的なref自体は**完全にサポートされ続けます**ので混同しないようにしてください）

Reviewer: ```suggestion
（String ref と異なり、一般的な ref 自体は**完全にサポートされ続けます**ので混同しないようにしてください）
```

</div> 
```

この `<div>` は `toggle` の値に関係なく常に背景色が赤になると思われるかもしれません。しかし、[このデモを見てわかるように](https://codesandbox.io/s/suspicious-sunset-g3jub)、`toggle` の値を `true` と `false` の間で切り替えると、背景色は `red` から始まり、その後 `transparent` と `blue` の間で交互に切り替わります。

> 

Reviewer: 余計な空行

  
**Reactは、スタイルルールの競合を検出し、警告をログに記録するようになりました。**この問題を修正するには、同一の CSS プロパティの短い記法と長い記法のバージョンを `style` プロパティ内で混在させないようにしてください。

### 一部の非推奨 string ref に関する注意点 {#warnings-for-some-deprecated-string-refs}

[String ref（文字列形式の ref）は古いレガシー API](/docs/refs-and-the-dom.html#legacy-api-string-refs)であり、既に勧められておらず、将来的に正式に非推奨となる予定です。

```js
<Button ref="myRef" />
```

(String ref と異なり、一般的な ref 自体は**完全にサポートされ続けます**ので混同しないようにしてください。）

> 

Reviewer: 余計な空行

</div> 
```

この `<div>` は `toggle` の値に関係なく常に背景色が赤になると思っていたかもしれません。しかし、[このデモを見てわかるように](https://codesandbox.io/s/suspicious-sunset-g3jub)、`toggle` の値を `true` と `false` の間で交互に変更すると、背景色は `red` から始まり、その後 `transparent` と `blue` の間で交互に変更されます。

  
**Reactは、スタイルルールの競合を検出し、警告をログに記録するようになりました。**
警告が検出されたソースコードを修正するには、同じCSSプロパティのshortandとlonghandのバージョンを`style`プロップに混在させないようにしてください。

> ### いくつかの非推奨文字列 refs に関する注意点{#warnings-for-some-deprecated-string-refs}

Reviewer: ```suggestion
### 一部の非推奨 string ref に関する注意点 {#warnings-for-some-deprecated-string-refs}
```

}
```

このようなコードはしばしばバグを示します。(ref は `ClassParent` で利用できると思われるかもしれませんが、代わりに `ClassWithRenderProp` に置かれてしまいます)。

> **このようなコードはお持ちでない可能性が高いでしょう**。もしそのようなコードがあって、それが意図的なものであれば、代わりに [`React.createRef()`](/docs/refs-and-the-dom.html#creating-refs) を使うよう変更してください:

Reviewer: ```suggestion
**このようなコードはお持ちでない可能性が高いでしょう**。もしそのようなコードがあって、それが意図的なものであれば、代わりに [`React.createRef()`](/docs/refs-and-the-dom.html#creating-refs) を使うよう変更してください：
```

}
```

このようなコードはしばしばバグを示します。(ref は `ClassParent` で利用できると思われるかもしれませんが、代わりに `ClassWithRenderProp` に置かれてしまいます)。

> **このようなコードはお持ちでない可能性が高いでしょう**。もしそのようなコードがあって、それが意図的なものであれば、代わりに [`React.createRef()`](/docs/refs-and-the-dom.html#creating-refs) を使うよう変更してください:

Reviewer: このあたりも空行の不統一

しかし、`unstable_createPortal` を採用している少数のライブラリが動作するように、`unstable_createPortal` をサポートされたエイリアスとして残していました。今後この不安定なエイリアスは非推奨とします。`unstable_createPortal` の代わりに `createPortal` を直接使ってください。これは全く同じシグネチャを持っています。

## その他の改良事項 {#other-improvements}

### hydration の警告におけるコンポーネントスタック {#component-stacks-in-hydration-warnings}

React は開発者向け警告にコンポーネントスタックを追加し、開発者がバグを特定してプログラムをデバッグできるようにしています。今回のリリースでは、以前はコンポーネントスタックが存在しなかった多くの開発者向け警告に、それが追加されました。例として、以前のバージョンにあったこのハイドレーション警告を考えてみましょう。

![A screenshot of the console warning, simply stating the nature of the hydration mismatch: "Warning: Expected server HTML to contain a matching div in div."](../images/blog/react-v16.13.0/hydration-warning-before.png)

コードとともにエラーを指摘していますが、どこにエラーが存在しているのか、次に何をすればいいのかが明確ではありません。このリリースでは、この警告にコンポーネントスタックが追加され、以下のようになりました。

![A screenshot of the console warning, stating the nature of the hydration mismatch, but also including a component stack : "Warning: Expected server HTML to contain a matching div in div, in div (at pages/index.js:4)..."](../images/blog/react-v16.13.0/hydration-warning-after.png)

これにより、問題がどこにあるのかが明確になり、より早くバグの場所を特定して修正することができます。

### 注目すべきバグ修正 {#notable-bugfixes}

このリリースには、他にもいくつかの注目すべき改善点が含まれています。

> - Strict 開発モードでは、React はライフサイクルメソッドを 2 回呼び出し、不要な副作用の可能性を洗い出すようにしています。このリリースでは、その動作を `shouldComponentUpdate` に追加しています。これは、`shouldComponentUpdate` に副作用がない限り、ほとんどのコードには影響しないはずです。これを修正するには、副作用のあるコードを `componentDidUpdate` に移動してください。

Reviewer 2: ここでいう副作用は、不要ではなく将来的に不具合の原因となる副作用という意味なので、`unwanted`の訳を変更してみたのですが如何でしょう？

```suggestion
- Strict 開発モードでは、React はライフサイクルメソッドを 2 回呼び出し、好ましくない副作用の可能性を洗い出すようにしています。このリリースでは、その動作を `shouldComponentUpdate` に追加しています。これは、`shouldComponentUpdate` に副作用がない限り、ほとんどのコードには影響しないはずです。これを修正するには、副作用のあるコードを `componentDidUpdate` に移動してください。
```

Author: ありがとうございます。こちらの文言に修正した方が良いと思います！

> 
> If you use Create React App or have the "react" preset with Babel 7+, you already have this plugin installed by default.

### `React.createFactory`の非推奨化{#deprecating-reactcreatefactory}

[`React.createFactory`](/docs/react-api.html#createfactory) は React 要素を作成するためのレガシーのヘルパです。このリリースでは、このメソッドに非推奨の警告が追加されています。これは将来のメジャーバージョンで削除される予定です。

`React.createFactory` の使用法を通常の JSX で置き換えてください。代わりに、この 1 行ヘルパをコピー＆ペーストするか、ライブラリとして公開することもできます。

```jsx
let createFactory = type => React.createElement.bind(null, type);
```

全く同じことをしています。

> ### `ReactDOM.unstable_createPortal` の非推奨化（`ReactDOM.createPortal` に変更を） {#deprecating-reactdomunstable_createportal-in-favor-of-reactdomcreateportal}

Reviewer: ```suggestion
### `ReactDOM.unstable_createPortal` の非推奨化、`ReactDOM.createPortal` に変更を {#deprecating-reactdomunstable_createportal-in-favor-of-reactdomcreateportal}
```

lint が文句言いそうなので表現を変えました、すみません

![A screenshot of the console warning, stating the nature of the hydration mismatch, but also including a component stack : "Warning: Expected server HTML to contain a matching div in div, in div (at pages/index.js:4)..."](../images/blog/react-v16.13.0/hydration-warning-after.png)

これにより、問題がどこにあるのかが明確になり、より早くバグの場所を特定して修正することができます。

### 注目すべきバグ修正 {#notable-bugfixes}

このリリースには、他にもいくつかの注目すべき改善点が含まれています。

- Strict 開発モードでは、React はライフサイクルメソッドを 2 回呼び出し、不要な副作用の可能性を洗い出すようにしています。このリリースでは、その動作を `shouldComponentUpdate` に追加しています。これは、`shouldComponentUpdate` に副作用がない限り、ほとんどのコードには影響しないはずです。これを修正するには、副作用のあるコードを `componentDidUpdate` に移動してください。

- Strict 開発モードでは、レガシーコンテクスト API の使用に関する警告には、警告のトリガとなったコンポーネントのスタックが含まれていませんでした。このリリースでは、欠けていたスタックが警告に追加されます。

- 無効な（disabled の）`<button>` 要素に対して `onMouseEnter` がトリガされないようになりました。

- ReactDOM は v16 を公開して以来、`version` のエクスポートがありませんでした。このリリースではそれが改めて追加されました。アプリケーションロジックでの使用はお勧めしませんが、同じページ上の ReactDOM のバージョン不一致や複数バージョンに関わる問題をデバッグする際に便利です。

これらの問題やその他の問題を解決してくれた貢献者に感謝します。完全な changelog は[下記](#changelog)を参照してください。

## インストール {#installation}

### React {#react}

React v16.13.0 はnpmで利用可能です。

> YarnでReact 16をインストールするには、下記を実行します:

Reviewer: ```suggestion
Yarn で React 16 をインストールするには、下記を実行します：
```

```bash
yarn add react@^16.13.0 react-dom@^16.13.0
```

> npmでReact 16をインストールするには、下記を実行します:

Reviewer: ```suggestion
npm で React 16 をインストールするには、下記を実行します：
```

```bash
npm install --save react@^16.13.0 react-dom@^16.13.0
```

> また、CDN経由でReactのUMDビルドも提供しています:

Reviewer: ```suggestion
また、CDN 経由で React の UMD ビルドも提供しています：
```

> 
> If you use Create React App or have the "react" preset with Babel 7+, you already have this plugin installed by default.

> ### `React.createFactory`は推奨しない{#deprecating-reactcreatefactory}

Reviewer: ```suggestion
### `React.createFactory` の非推奨化 {#deprecating-reactcreatefactory}
```

[`React.createFactory`](/docs/react-api.html#createfactory) は React 要素を作成するためのレガシーのヘルパです。このリリースでは、このメソッドに非推奨の警告が追加されています。これは将来のメジャーバージョンで削除される予定です。

`React.createFactory` の使用法を通常の JSX で置き換えてください。代わりに、この 1 行ヘルパをコピー＆ペーストするか、ライブラリとして公開することもできます。

```jsx
let createFactory = type => React.createElement.bind(null, type);
```

全く同じことをしています。

> ### `ReactDOM.unstable_createPortal`を`ReactDOM.createPortal`に変更することを推奨しない{#deprecating-reactdomunstable_createportal-in-favor-of-reactdomcreateportal}

Reviewer: ```suggestion
### `ReactDOM.unstable_createPortal` の非推奨化（`ReactDOM.createPortal` に変更を） {#deprecating-reactdomunstable_createportal-in-favor-of-reactdomcreateportal}
```
意味が逆です。「AをBにすることを非推奨化」ではなく「Aを非推奨化して代わりにBを使う」です。

![A screenshot of the console warning, stating the nature of the hydration mismatch, but also including a component stack : "Warning: Expected server HTML to contain a matching div in div, in div (at pages/index.js:4)..."](../images/blog/react-v16.13.0/hydration-warning-after.png)

これにより、問題がどこにあるのかが明確になり、より早くバグの場所を特定して修正することができます。

### 注目すべきバグ修正 {#notable-bugfixes}

このリリースには、他にもいくつかの注目すべき改善点が含まれています。

- Strict 開発モードでは、React はライフサイクルメソッドを 2 回呼び出し、不要な副作用の可能性を洗い出すようにしています。このリリースでは、その動作を `shouldComponentUpdate` に追加しています。これは、`shouldComponentUpdate` に副作用がない限り、ほとんどのコードには影響しないはずです。これを修正するには、副作用のあるコードを `componentDidUpdate` に移動してください。

- Strict 開発モードでは、レガシーコンテクスト API の使用に関する警告には、警告のトリガとなったコンポーネントのスタックが含まれていませんでした。このリリースでは、欠けていたスタックが警告に追加されます。

- 無効な（disabled の）`<button>` 要素に対して `onMouseEnter` がトリガされないようになりました。

- ReactDOM は v16 を公開して以来、`version` のエクスポートがありませんでした。このリリースではそれが改めて追加されました。アプリケーションロジックでの使用はお勧めしませんが、同じページ上の ReactDOM のバージョン不一致や複数バージョンに関わる問題をデバッグする際に便利です。

これらの問題やその他の問題を解決してくれた貢献者に感謝します。完全な changelog は[下記](#changelog)を参照してください。

## インストール {#installation}

### React {#react}

React v16.13.0 はnpmで利用可能です。

> YarnでReact 16をインストールするには、下記を実行します

Reviewer: ```suggestion
YarnでReact 16をインストールするには、下記を実行します：
```

```bash
yarn add react@^16.13.0 react-dom@^16.13.0
```

> npmでReact 16をインストールするには、下記を実行します

Reviewer: ```suggestion
npmでReact 16をインストールするには、下記を実行します：
```

```bash
npm install --save react@^16.13.0 react-dom@^16.13.0
```

> また、CDN経由でReactのUMDビルドも提供しています

Reviewer: ```suggestion
また、CDN経由でReactのUMDビルドも提供しています：
```

### Deprecating `React.createFactory` {#deprecating-reactcreatefactory}

[`React.createFactory`](/docs/react-api.html#createfactory) はReact要素を作成するための旧版のルパーです。このリリースでは、メソッドに非推奨の警告が追加されています。これは将来のメジャーバージョンで削除される予定です。

`React.createFactory` の使用法を通常の JSX で置き換えてください。代わりに、この一行ヘルパーをコピー＆ペーストするか、ライブラリとして公開することもできます。

```jsx
let createFactory = type => React.createElement.bind(null, type);
```

全く同じことをしています。

### Deprecating `ReactDOM.unstable_createPortal` in favor of `ReactDOM.createPortal` {#deprecating-reactdomunstable_createportal-in-favor-of-reactdomcreateportal}

React 16がリリースされたとき、`createPortal`は公式にサポートされるAPIになりました。

> しかし、`unstable_createPortal`を採用している少数のライブラリが動作するように、`unstable_createPortal`をサポートされたエイリアスとして残していました。現在、この不安定なエイリアスは非推奨となっています。`unstable_createPortal` の代わりに `createPortal` を直接使ってください。これは全く同じシグネチャを持っています。

Reviewer: ```suggestion
しかし、`unstable_createPortal`を採用している少数のライブラリが動作するように、`unstable_createPortal`をサポートされたエイリアスとして残していました。今後この不安定なエイリアスは非推奨とします。`unstable_createPortal` の代わりに `createPortal` を直接使ってください。これは全く同じシグネチャを持っています。
```

この We are now deprecating はこれからすることだと思います

Reviewer: まだ suggestion が反映されていません

### Notable bugfixes {#notable-bugfixes}

このリリースには、他にもいくつかの注目すべき改善点が含まれています。

- 厳密な開発モードでは、React はライフサイクルメソッドを2回呼び出し、不要な副作用の可能性を洗い出すようにしています。このリリースでは、その動作を `shouldComponentUpdate` に追加しています。これは、`shouldComponentUpdate`に副作用がない限り、ほとんどのコードには影響しないはずです。これを修正するには、副作用のあるコードを `componentDidUpdate` に移動してください。

- 厳密な開発モードでは、レガシーコンテキストAPIの使用に関する警告には、警告のトリガーとなったコンポーネントのスタックが含まれていませんでした。このリリースでは、不足していたスタックが警告に追加されます。

- 無効な `<button>` 要素に対して `onMouseEnter` がトリガーされないようになりました。

- ReactDOM は v16 を公開して以来、`version` のエクスポートがありませんでした。このリリースではそれが追加されました。アプリケーションロジックでの使用はお勧めしませんが、同じページ上のReactDOMのバージョンの不一致や複数のバージョンの問題をデバッグする際に便利です。

これらの問題やその他の問題を解決してくれた貢献者に感謝します。完全なchangelog [下記](#changelog)を参照してください。

## Installation {#installation}

### React {#react}

React v16.13.0 はnpmで利用可能です。

YarnでReact 16をインストールするには、下記を実行します

```bash
yarn add react@^16.13.0 react-dom@^16.13.0
```

npmでReact 16をインストールするには、下記を実行します

```bash
npm install --save react@^16.13.0 react-dom@^16.13.0
```

また、CDN経由でReactのUMDビルドも提供しています

```html
<script crossorigin src="https://unpkg.com/react@16/umd/react.production.min.js"></script>
<script crossorigin src="https://unpkg.com/react-dom@16/umd/react-dom.production.min.js"></script>
```

> [詳細のスクリプト設置方法](/docs/installation.html)については、説明書を参照してください。

Reviewer: まだ suggestion が反映されていません
